BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch

```diff
@@ -0,0 +1,258 @@
++            int inv = listenerClass.isInterface() ? Opcodes.INVOKEINTERFACE : Opcodes.INVOKEVIRTUAL;
++            mv.visitMethodInsn(inv, Type.getInternalName(listenerClass), handlerMethod.getName(), Type.getMethodDescriptor(handlerMethod), listenerClass.isInterface());
++            if(handlerMethod.getReturnType() != void.class) {
++                mv.visitInsn(Opcodes.POP);
++            }
++            mv.visitInsn(Opcodes.RETURN);
++            mv.visitMaxs(0, 0);
++            mv.visitEnd();
++        }
++
++        cw.visitEnd();
++        return cw.toByteArray();
++    }
++
++    private static Class<?> defineClass(ClassLoader loader, String name, byte[] data) {
++        try {
++            return (Class<?>) defineClassMethod.invoke(loader, name, data, 0, data.length);
++        } catch (Exception e) {
++            throw new RuntimeException("Failed to define class '" + name + "'", e);
++        }
++    }
++
++    static {
++        try {
++            defineClassMethod = ClassLoader.class.getDeclaredMethod("defineClass", String.class, byte[].class, int.class, int.class);
++            defineClassMethod.setAccessible(true);
++        } catch (Exception e) {
++            throw new Error("Failed to get ClassLoader#defineClass method", e);
++        }
++    }
++}
+\ No newline at end of file
+diff --git a/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/EventExecutor.java b/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/EventExecutor.java
+new file mode 100644
+index 00000000..9f8aac30
+--- /dev/null
++++ b/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/EventExecutor.java
+@@ -0,0 +1,20 @@
++package io.github.waterfallmc.waterfall.impl.event;
++
++import java.lang.reflect.Method;
++import java.lang.reflect.Modifier;
++
++/**
++ * @author Mark Vainomaa
++ */
++public interface EventExecutor {
++    void execute(Object listener, Object event);
++
++    static EventExecutor create(Method method) {
++        if (Modifier.isPublic(method.getModifiers())) {
++            return ASMEventExecutor.getEventExecutor(method);
++        } else {
++            method.setAccessible(true);
++            return new MethodHandleEventExecutor(method);
++        }
++    }
++}
+diff --git a/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/MethodHandleEventExecutor.java b/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/MethodHandleEventExecutor.java
+new file mode 100644
+index 00000000..844cf53d
+--- /dev/null
++++ b/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/MethodHandleEventExecutor.java
+@@ -0,0 +1,44 @@
++package io.github.waterfallmc.waterfall.impl.event;
++
++import lombok.SneakyThrows;
++
++import java.lang.invoke.MethodHandle;
++import java.lang.invoke.MethodHandles;
++import java.lang.reflect.Method;
++import java.lang.reflect.Modifier;
++
++/**
++ * @author Mark Vainomaa
++ */
++class MethodHandleEventExecutor implements EventExecutor {
```

Comment:
i think there were huge discussions about method handle vs normal reflection and i heard that methodhandles are in the end slower than normal reflection, as the jvm optimizes normal reflection also to asm at some point

Reply:
Yeeah maybe indeed.

Reply:
In case of Bungee not executing non-public methods (either intentionally or not intentionally), then this event executor implementation becomes irrelevant.